### PR TITLE
ci(smart-contracts): run smart contracts coverage task on CI

### LINKED ISSUE #7739: Reporting for smart contracts coverage
Coverage has been added in #7635 but we still lack integration into CI and proper reporting to evaluate what is missing. Per previous discussions, the coverage target for the main contracts (such as publicLock or even UDT) should be close to 100%. 

We are in the process of upgrading UDT and integrate contracts "integration" testing on a mainnet fork (see #7655 ), we could as well add coverage repotring at least for the unit testing part. Could use Coveralls or another similar service

Sidenote: maybe #1781 can be closed?

### PR DESCRIPTION
**Pull request by clemsos**
# Description

Run smart contracts coverage on CI

# Issues

Fixes #7739
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by clemsos** [comment]
because `solidty-coverage` currently has issues regarding block mining in tests on hardhat (see https://github.com/sc-forks/solidity-coverage/issues/707), I skipped the tests that were requiring EIP-1559. 

Not ideal but that should get us a green build for coverage for now

**Comment by clemsos** [comment]
Results from the CI

File                           |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-------------------------------|----------|----------|----------|----------|----------------|
 contracts/                    |    80.14 |       58 |     90.7 |    80.14 |                |
  PublicLock.sol               |      100 |      100 |      100 |      100 |                |
  Unlock.sol                   |    75.49 |    59.52 |    91.67 |    75.49 |... 450,451,493 |
  UnlockDiscountTokenV3.sol    |    71.43 |       50 |      100 |    71.43 |          26,31 |
  UnlockProtocolGovernor.sol   |    92.59 |       50 |    85.71 |    92.59 |        113,131 |
  UnlockProtocolTimelock.sol   |      100 |      100 |      100 |      100 |                |
 contracts/hooks/              |    91.67 |       90 |      100 |      100 |                |
  ERC721BalanceOfHook.sol      |    91.67 |       90 |      100 |      100 |                |
 contracts/interfaces/         |      100 |      100 |      100 |      100 |                |
  IMintableERC20.sol           |      100 |      100 |      100 |      100 |                |
  IPublicLock.sol              |      100 |      100 |      100 |      100 |                |
  IUnlock.sol                  |      100 |      100 |      100 |      100 |                |
 contracts/interfaces/hooks/   |      100 |      100 |      100 |      100 |                |
  ILockKeyCancelHook.sol       |      100 |      100 |      100 |      100 |                |
  ILockKeyPurchaseHook.sol     |      100 |      100 |      100 |      100 |                |
  ILockTokenURIHook.sol        |      100 |      100 |      100 |      100 |                |
  ILockValidKeyHook.sol        |      100 |      100 |      100 |      100 |                |
 contracts/mixins/             |    99.27 |    93.69 |    98.89 |    99.28 |                |
  MixinConvenienceOwnable.sol  |      100 |      100 |      100 |      100 |                |
  MixinDisable.sol             |      100 |      100 |      100 |      100 |                |
  MixinERC721Enumerable.sol    |      100 |      100 |      100 |      100 |                |
  MixinFunds.sol               |      100 |    83.33 |      100 |      100 |                |
  MixinGrantKeys.sol           |      100 |      100 |      100 |      100 |                |
  MixinKeys.sol                |     99.2 |    95.31 |    96.77 |    99.23 |            118 |
  MixinLockCore.sol            |      100 |       95 |      100 |      100 |                |
  MixinLockMetadata.sol        |      100 |      100 |      100 |      100 |                |
  MixinPurchase.sol            |      100 |    91.67 |      100 |      100 |                |
  MixinRefunds.sol             |      100 |      100 |      100 |      100 |                |
  MixinRoles.sol               |      100 |    66.67 |      100 |      100 |                |
  MixinTransfer.sol            |    96.92 |    94.12 |      100 |    96.97 |        268,303 |
 contracts/utils/              |    82.14 |    57.14 |       60 |       80 |                |
  Clone2Factory.sol            |        0 |        0 |        0 |        0 |          27,70 |
  LockSerializer.sol           |      100 |      100 |      100 |      100 |                |
  UnlockContextUpgradeable.sol |    33.33 |      100 |       25 |    33.33 |          19,29 |
  UnlockInitializable.sol      |    88.89 |       75 |    66.67 |    81.82 |          73,74 |
  UnlockOwnable.sol            |    45.45 |       50 |    57.14 |       50 |... 74,81,82,83 |
-------------------------------|----------|----------|----------|----------|----------------|
All files                      |    93.14 |    85.36 |    91.61 |    93.07 |                |
-------------------------------|----------|----------|----------|----------|----------------|

**Comment by clemsos** [comment]
Note that Unlock coverage appears low are we are skipping all the `recordKeyPurchase` logic due to the bug in the lib

**Comment by julien51** [comment]
Oh interesting! What's the bug?

I think we can ignore `UnlockDiscountTokenV3.sol`, `UnlockProtocolGovernor.sol`, `Clone2Factory.sol` (what's this again?).

And then what are `UnlockContextUpgradeable.sol`, `UnlockInitializable.sol` and `UnlockOwnable.sol`?

Let's make sure we add `MixinKeys.sol` and `MixinTransfer.sol`?

**Comment by clemsos** [comment]
> And then what are `UnlockContextUpgradeable.sol`, `UnlockInitializable.sol` and `UnlockOwnable.sol`?

These are flattened version of OZ contracts we checked in for backwards compat

**Comment by julien51** [comment]
Ok, LG. Let's then just add tests to make sure we cover Unlock.sol, MixinKeys.sol and MixinTransfer.sol?

**Comment by clemsos** [comment]
updated

File                          |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
------------------------------|----------|----------|----------|----------|----------------|
 contracts/                   |    96.43 |    80.95 |     96.3 |    96.43 |                |
  PublicLock.sol              |      100 |      100 |      100 |      100 |                |
  Unlock.sol                  |    96.08 |    80.95 |    95.83 |    96.08 |408,412,433,493 |
  UnlockProtocolTimelock.sol  |      100 |      100 |      100 |      100 |                |
 contracts/hooks/             |    92.59 |    90.91 |      100 |      100 |                |
  ERC20BalanceOfHook.sol      |    93.33 |    91.67 |      100 |      100 |                |
  ERC721BalanceOfHook.sol     |    91.67 |       90 |      100 |      100 |                |
 contracts/interfaces/        |      100 |      100 |      100 |      100 |                |
  IMintableERC20.sol          |      100 |      100 |      100 |      100 |                |
  IPublicLock.sol             |      100 |      100 |      100 |      100 |                |
  IUnlock.sol                 |      100 |      100 |      100 |      100 |                |
 contracts/interfaces/hooks/  |      100 |      100 |      100 |      100 |                |
  ILockKeyCancelHook.sol      |      100 |      100 |      100 |      100 |                |
  ILockKeyPurchaseHook.sol    |      100 |      100 |      100 |      100 |                |
  ILockTokenURIHook.sol       |      100 |      100 |      100 |      100 |                |
  ILockValidKeyHook.sol       |      100 |      100 |      100 |      100 |                |
 contracts/mixins/            |    99.27 |    93.69 |    98.89 |    99.28 |                |
  MixinConvenienceOwnable.sol |      100 |      100 |      100 |      100 |                |
  MixinDisable.sol            |      100 |      100 |      100 |      100 |                |
  MixinERC721Enumerable.sol   |      100 |      100 |      100 |      100 |                |
  MixinFunds.sol              |      100 |    83.33 |      100 |      100 |                |
  MixinGrantKeys.sol          |      100 |      100 |      100 |      100 |                |
  MixinKeys.sol               |     99.2 |    95.31 |    96.77 |    99.23 |            118 |
  MixinLockCore.sol           |      100 |       95 |      100 |      100 |                |
  MixinLockMetadata.sol       |      100 |      100 |      100 |      100 |                |
  MixinPurchase.sol           |      100 |    91.67 |      100 |      100 |                |
  MixinRefunds.sol            |      100 |      100 |      100 |      100 |                |
  MixinRoles.sol              |      100 |    66.67 |      100 |      100 |                |
  MixinTransfer.sol           |    96.92 |    94.12 |      100 |    96.97 |        268,303 |
------------------------------|----------|----------|----------|----------|----------------|
All files                     |    98.37 |    91.48 |    98.35 |    98.74 |                |
------------------------------|----------|----------|----------|----------|----------------|

**Comment by clemsos** [comment]
> Ok, LG. Let's then just add tests to make sure we cover Unlock.sol, MixinKeys.sol and MixinTransfer.sol?

for Unlock.sol, these are the functions base on `basefee` that we have to ignore because of solc-coverage not supporting EIP-1559

On MixinKeys.sol, there was an unused internal function - deleted in https://github.com/unlock-protocol/unlock/pull/8844

On MixinTransfer, there is one line in `getTransferFee` and one in `_checkOnERC721Received` that are not covered

File: smart-contracts/test/UnlockDiscountToken/transferingTokens.js [hunk 1]
```diff
@@ -12,13 +15,16 @@ let unlock
 let udt
 let lock
 
+// skip on coverage until solidity-coverage supports EIP-1559
+const describeOrskip = process.env.IS_COVERAGE ? describe.skip : describe
```
Comment (julien51) [inline smart-contracts/test/UnlockDiscountToken/transferingTokens.js:19]:
```suggestion
const describeOrSkip = process.env.IS_COVERAGE ? describe.skip : describe
```

File: smart-contracts/test/UnlockDiscountToken/transferingTokens.js [hunk 2]
```diff
@@ -218,8 +228,8 @@ contract('UnlockDiscountToken (l2/sidechain) / granting Tokens', (accounts) => {
     })
   })
 
-  describe('grant capped by % growth', () => {
-    beforeEach(async () => {
+  describeOrskip('grant capped by % growth', () => {
```
Comment (julien51) [inline smart-contracts/test/UnlockDiscountToken/transferingTokens.js:231]:
```suggestion
  describeOrSkip('grant capped by % growth', () => {
```

File: smart-contracts/test/UnlockDiscountToken/upgrades.js [hunk 3]
```diff
@@ -10,6 +10,9 @@ const createLockHash = require('../helpers/createLockCalldata')
 
 const Locks = require('../fixtures/locks')
 
+// skip on coverage until solidity-coverage supports EIP-1559
+const describeOrskip = process.env.IS_COVERAGE ? describe.skip : describe
```
Comment (julien51) [inline smart-contracts/test/UnlockDiscountToken/upgrades.js:14]:
```suggestion
const describeOrSkip = process.env.IS_COVERAGE ? describe.skip : describe
```

File: smart-contracts/test/UnlockDiscountToken/upgrades.js [hunk 4]
```diff
@@ -320,8 +327,9 @@ contract('UnlockDiscountToken upgrade', async () => {
       })
     })
 
-    describe('mint capped by % growth', () => {
-      beforeEach(async () => {
+    describeOrskip('mint capped by % growth', () => {
```
Comment (julien51) [inline smart-contracts/test/UnlockDiscountToken/upgrades.js:330]:
```suggestion
    describeOrSkip('mint capped by % growth', () => {
```

**Review by julien51 — APPROVED** [review]
Small nits!

**Pull request merged**